src/te/autodiff/ad_simplify.cc

```diff
@@ -0,0 +1,1305 @@
+  return RemoveRedundantInequalitiesMutator(known)(expr);
+}
+
+// Extract the given expr under the given condition as a separate tensor if the volume of the
+// extracted tensor will be less than the volume of the outer_axis
+PrimExpr TrySimplifyCompute(const PrimExpr& expr, const PrimExpr& cond,
+                            const Array<Var>& outer_axis, const Map<Var, Range>& vranges) {
+  // solve cond, e.g., (jac_i0 == i) && (jac_i1 == j)
+  arith::IntConstraints domain_to_solve(outer_axis, vranges,
+                                        FactorOutAtomicFormulas(cond).to_array());
+  auto res = SimplifyDomain(domain_to_solve);
+
+  arith::Analyzer analyzer;
+  analyzer.Bind(res->dst->ranges);
+  PrimExpr new_expr = analyzer.Simplify(Substitute(expr, res->src_to_dst),
+                                        ARITH_SIMPLIFY_REWRITE_CANONICAL_REWRITE);
+  // TODO(yzhliu): This is mostly done to simplify if_then_else
+  // which is not realized by the canonical simplifier
+  new_expr = RemoveRedundantInequalities(new_expr, res->dst->relations);
+
+  // Keep only those variables of the new vars which are used in the new_expr
+  Array<Var> used_res_variables;
+  for (const Var& var : res->dst->variables) {
+    if (ExprUseVar(new_expr, var)) {
+      CHECK(res->dst->ranges.count(var)) << "Range of " << var << " cannot be inferred.";
+      used_res_variables.push_back(var);
+    }
+  }
+
+  // If the expression does not use vars then it is probably better to keep it inlined
+  if (used_res_variables.empty()) {
+    // We can return the new_expr here instead of the old expr because it doesn't use variables
+    // otherwise we would need to replace the new vars or create a let-expression
+    return new_expr;
+  }
+
+  // If it's already tensor[...] then it will probably be useless to further simplify it.
+  if (new_expr.as<ProducerLoadNode>()) {
+    return expr;
+  }
+
+  // Compute volumes before and after
+  PrimExpr old_volume = make_const(DataType::Int(64), 1);
+  for (const Var& var : outer_axis) {
+    CHECK(vranges.count(var)) << "Range of " << var << " was not provided.";
+    old_volume = old_volume * vranges[var]->extent;
+  }
+
+  PrimExpr new_volume = make_const(DataType::Int(64), 1);
+  for (const Var& var : used_res_variables) {
+    new_volume = new_volume * res->dst->ranges[var]->extent;
+  }
+
+  // if we can prove that the old volume is not greater than the new volume then
+  // prefer the old expression.
+  arith::Analyzer ana_vranges;
+  ana_vranges.Bind(vranges);
+  if (ana_vranges.CanProve(old_volume <= new_volume)) {
+    return expr;
+  }
+
+  Tensor tensor = TensorFromExpr(new_expr, IterVarsFromMap(used_res_variables, res->dst->ranges),
+                                 "extracted_tensor");
+
+  Array<PrimExpr> args;
+  for (const Var& var : used_res_variables) {
+    args.push_back(res->dst_to_src[var]);
+  }
+
+  return ProducerLoad(tensor, args);
+}
+
+class FreeVarsVisitor : public StmtExprVisitor {
```

Comment:
move this to common file.

Reply:
I removed it and use `UndefinedVars` instead now.